### DISCUSSION TITLE
Adding virtual method to the component class

### DISCUSSION THREAD

**Discussion by Abbas111977** [SpiceSharp/SpiceSharp#195] [Q&A] [answered]
Dears,
I am trying to add the following virtual methods to component class then i have override it in derived class but it seems that it give me an exception when i try to access it , can you please help me with this issue ? 
i added the following code at the end of the component.cs class

        /// <inheritdoc/>
        public virtual string GetEquivalentSpiceLine()
        {
            return null;
        }
        /// <inheritdoc/>
        public virtual XElement GetEquivelentXml()
        {
            return null;
        }

and in the derived class i added the following code:

        /// <inheritdoc/>
        public override string GetEquivalentSpiceLine()
        {
            return "";
        }
        /// <inheritdoc/>
        public override XElement GetEquivelentXml()
        {
            return null;
        }

**Comment by svenboulanger** [comment, marked as answer]
That's not much information to go on. You mention an exception but there is no exception in your example code. Are you trying to modify the Spice# library? How are you trying to access those methods?

**Reply by Abbas111977** [reply]
Yes, i am trying to customize the library, and the exception was MissingMethodException when i try to call the above methods.

**Reply by svenboulanger** [reply]
That means you're trying to dynamically call those methods and you're probably not accessing the right type. I advise trying to test by accessing those methods at compile time before trying to access them dynamically.

**Reply by Abbas111977** [reply]
I am adding these methods statically ( just adding the code to the existing classes ) and it dose not complain at compile time, but in runtime it show me the above mentioned exception !!!

**Reply by svenboulanger** [reply]
With the information you have provided, I believe it is impossible to find out where the issue is. The [MissingMethodException](https://docs.microsoft.com/en-us/dotnet/api/system.missingmethodexception?view=net-5.0) already cleared up that you are trying to access the method dynamically, and that the method does not exist. What I believe is going wrong, is that whatever object you are accessing is not of the `Component` type.

**Reply by Abbas111977** [reply]
put it easily, I am trying to add new property or method to the Component class so that I can access it form all the derived classes ( Resistor, Diode, ... etc. ) and my own sub derived classes ( MyResistor:Resistor)  , is it possible ?
example:
Resistor r = new Resistor("r", 1, 2, 10);
r.DoSomeThing();
MyResistor r1 = new MyResistor("r1", 3, 4, 30);// MyResistor is derived from Resistor class
r1.DoSomeThing();
 
and I have another question please , is it possible to change the nodes of a component after creating it ?

**Reply by svenboulanger** [reply]
If you are cloning the repository, modify the `Component` class and compile the project yourself, then yes that should be possible.
If you are using the NuGet package, and try to reimplement the `Component` class, then no that is not possible. You have to compile it yourself.

> is it possible to change the nodes of a component after creating it ?

Yes, there should be a `Connect(...)` method for components (`IComponent`) interface. You cannot change the connections of behaviors though (i.e. what is used by the simulation once it starts running).

**Reply by Abbas111977** [reply]
I have tried already to compile the Component class and its compiled without any complains but when i call the added method it shows me the exception
![x3](https://user-images.githubusercontent.com/62871714/138512204-42098f4f-7046-45a7-bda8-d190ae1fd635.jpg)
![x2](https://user-images.githubusercontent.com/62871714/138512207-b34f761b-a565-45ea-9ca6-aa26373739f6.jpg)
![x1](https://user-images.githubusercontent.com/62871714/138512212-739d226d-3cf8-40f9-a39d-ddc23c12cc79.png)

**Reply by svenboulanger** [reply]
The only thing that comes to mind is what [this person on StackOverflow](https://stackoverflow.com/questions/7578583/method-not-found-on-runtime) described.

If you compile and run the application, but you accidentally have that exe use the old Spice# DLL, then this would explain the exception. Try removing the DLL from your build output, double check that you are referencing your own project and not a NuGet package somewhere.

**Reply by Abbas111977** [reply]
That solve it, Many thanks.
